int32[] ip
float32[] x
float32[] y
float32[] theta
